keps/1282-requeue-strategy/README.md

```diff
@@ -0,0 +1,306 @@
+	objB := b.(*workload.Info)
+	p1 := utilpriority.Priority(objA.Obj)
+	p2 := utilpriority.Priority(objB.Obj)
+
+	if p1 != p2 {
+		return p1 > p2
+	}
+
+	tA := workload.GetQueueOrderTimestamp(objA.Obj)
+	tB := workload.GetQueueOrderTimestamp(objB.Obj)
+	return !tB.Before(tA)
+}
+```
+
+```go
+package workload
+
+// ...
+
+func GetQueueOrderTimestamp(w *kueue.Workload) *metav1.Time {
+	if c := apimeta.FindStatusCondition(w.Status.Conditions, kueue.WorkloadEvicted); c != nil && c.Status == metav1.ConditionTrue && c.Reason == kueue.WorkloadEvictedByPodsReadyTimeout {
+		return &c.LastTransitionTime
+	}
+	return &w.CreationTimestamp
+}
+```
+
+#### Proposed Sorting
+
+Add types that contain information about queue ordering strategy to the `apis/config/<version>` package. For example:
+
+```go
+package v1beta1
+
+// ...
+
+type QueueOrderingCriteria string
+
+const (
+  OrderOnEvictionTimestamp = QueueOrderingCriteria("UseEvictionTimestamp")
+  OrderOnCreationTimestamp = QueueOrderingCriteria("UseCreationTimestamp")
+)
+
+type QueueOrderingStrategy struct{
+  RequeueOnPriorityPreemption QueueOrderingCriteria
+  RequeueOnPodsReadyTimeout   QueueOrderingCriteria
+}
+```
+
+The `pkg/workload` package could have the `GetQueueOrderTimestamp` function modified to include conditionals that control which timestamp to return based on the configured ordering strategy.
+
+```go
+func GetQueueOrderTimestamp(w *kueue.Workload, s QueueOrderingStrategy) *metav1.Time {
+  switch getEvictionReason(w) {
+    case kueue.WorkloadEvictedByPodsReadyTimeout:
+      if s.RequeueOnPodsReadyTimeout == OrderOnEvictionTimestamp {
+        return &c.LastTransitionTime
+      } else {
+        return &c.CreationTimestamp
+      }
+    case kueue.WorkloadEvictedByPreemption:
+      if s.RequeueOnPriorityPreemption == OrderOnEvictionTimestamp {
+        return &c.LastTransitionTime
+      } else {
+        return &c.CreationTimestamp
+      }
+    default:
+      return &w.CreationTimestamp
+  }
+}
+
+// ...
+
+func getEvictionReason(w *kueue.Workload) string {
+	if c := apimeta.FindStatusCondition(w.Status.Conditions, kueue.WorkloadEvicted); c != nil && c.Status == metav1.ConditionTrue {
+    return c.Reason
+  }
+  return ""
+}
+```
+
+#### Preemption/Preemptor Flapping
+
+To address the [following concern](https://github.com/kubernetes-sigs/kueue/issues/1282#issue-1966106166):
+
+> The case of preemption might be more tricky: we want the preempted jobs to be readmitted as soon as possible. But if a job is waiting for more than one job to be preempted and the queueing strategy is BestEffortFIFO, we don't want the preempted pods to take the head of the queue.
+> Maybe we need to hold them until the preemptor job is admitted, and then they should use the regular priority.
+
+The `pkg/queue` package could have the existing `queueOrderingFunc()` modified to add sorting based on who is the preemptor (might need to add a condition to the Workload to track this).
```

Comment:
It looks like my concern was unfounded.
We handle this problem here:
https://github.com/kubernetes-sigs/kueue/blob/c2f21f4eaeb9393ab117cf836b90ca4b7b6951be/pkg/queue/cluster_queue_best_effort_fifo.go#L43

Basically, if the workload has preempted any workloads, it's put back in the queue immediately.

Since nothing changes about it and it will have a higher priority than the preempted workloads, it should be at the head of the queue.